controllers/wordpressinstance_controller.go

```diff
@@ -50,283 +59,90 @@ type WordpressInstanceReconciler struct {
+		return errors.Wrap(err, fmt.Sprintf("cannot create %s/%s in namespace %s", kubernetesApplication.Kind, kubernetesApplication.GetName(), kubernetesApplication.GetNamespace()))
 	}
-
 	return nil
 }
 
-func (r *WordpressInstanceReconciler) createOutputObject(ctx context.Context, obj *unstructured.Unstructured, i *wordpressv1alpha1.WordpressInstance) error {
-	// if we decoded a non-nil unstructured object, try to create it now
-	if obj == nil {
-		return nil
+func GetEndpoint(ctx context.Context, kube client.Client, wp wordpressv1alpha1.WordpressInstance) (string, error) {
+	serviceResource := &v1.Service{}
+	key, err := GetServiceResourceObjectKey(wp)
+	if err != nil {
+		return "", err
 	}
-
-	// set an owner reference on the object
-	obj.SetOwnerReferences([]metav1.OwnerReference{
-		AsOwner(ReferenceTo(i, wordpressv1alpha1.WordpressInstanceGroupVersionKind)),
-	})
-
-	r.Log.V(1).Info(
-		"creating object",
-		"name", obj.GetName(),
-		"namespace", obj.GetNamespace(),
-		"apiVersion", obj.GetAPIVersion(),
-		"kind", obj.GetKind(),
-		"ownerRefs", obj.GetOwnerReferences())
-
-	if err := r.Client.Create(ctx, obj); err != nil && !kerrors.IsAlreadyExists(err) {
-		return errors.Wrapf(err, "failed to create object %s: %s", obj.GetName(), err)
+	if err := kube.Get(ctx, key, serviceResource); err != nil {
+		return "", err
 	}
-
-	return nil
+	ingress := serviceResource.Status.LoadBalancer.Ingress[0]
+	if ingress.IP != "" {
+		return ingress.IP, nil
+	}
+	if ingress.Hostname != "" {
+		return ingress.Hostname, nil
+	}
+	return "", nil
 }
 
-func (r *WordpressInstanceReconciler) SetupWithManager(mgr ctrl.Manager) error {
-	return ctrl.NewControllerManagedBy(mgr).
-		For(&wordpressv1alpha1.WordpressInstance{}).
-		Complete(r)
+func GetServiceResourceObjectKey(wp wordpressv1alpha1.WordpressInstance) (client.ObjectKey, error) {
+	kubernetesApplication := GetKubernetesApplication(wp, "")
+	for _, template := range kubernetesApplication.Spec.ResourceTemplates {
+		if template.Spec.Template.GetKind() == "Service" {
+			return client.ObjectKey{Name: template.Spec.Template.GetName(), Namespace: template.Spec.Template.GetNamespace()}, nil
+		}
+	}
+	return client.ObjectKey{}, fmt.Errorf("cannot find service type kubernetes application resource for %s/%s in namespace %s", wp.Kind, wp.GetName(), wp.GetNamespace())
 }
 
-// HACK:
-// The utility methods below have been copied from Crossplane (https://github.com/crossplaneio/crossplane)
-
-// ReferenceTo returns an object reference to the supplied object, presumed to
-// be of the supplied group, version, and kind.
-func ReferenceTo(o metav1.Object, of schema.GroupVersionKind) *corev1.ObjectReference {
-	v, k := of.ToAPIVersionAndKind()
-	return &corev1.ObjectReference{
-		APIVersion: v,
-		Kind:       k,
-		Namespace:  o.GetNamespace(),
-		Name:       o.GetName(),
-		UID:        o.GetUID(),
+func FillWithDefaults(wp *wordpressv1alpha1.WordpressInstance) {
```

Comment:
I'm not super up on the go patterns for doing this, but I would normally prefer to see a default-setting method as taking the defaults as a parameter, then merging the defaults into the object. The behavior would be similar to a deep copy into an existing object. Is that something that is easy and idiomatic in go? One nice benefit is that the function is stateless if the default object is passed in. Another nice benefit is that it's easy to add more defaults by just adding them to the default object.

Reply:
> DeepCopyInto assigns the values even if a value is already there. So, it wouldn't work for defaulting.

Does it assign the values if the `from` object has empty values? It [seems like it does not](https://github.com/crossplaneio/sample-stack-wordpress/pull/17/files#diff-a9ceb1ef9a800df85f5f80f38e4b6deeR89)

I was talking about the approach of starting with a third empty object. Then layer the default fields in, then layer the original object's fields in. Here is a more complete example of what I was imagining:

```go
// defaults for a wordpress instance
const defaultWp = WordpressInstance{... default params ...}

var clientWp = // this is fetched from the client.Get call
var wpInstance = FillWithDefaults(&clientWp, &defaultWp)

// then, use wpInstance
// ...

// meanwhile . . .
func FillWithDefaults(wp *wordpressv1alpha1.WordpressInstance, defaultWp *wordpressv1alpha1.WordpressInstance)  wordpressv1alpha1.WordpressInstance {
  // this starts off empty, but later stores the final merged result
  var resultWp = WordpressInstance{}

  defaultWp.DeepCopyInto(&resultWp)
  wp.DeepCopyInto(&resultWp)

  return resultWp
}
```

Reply:
> Does it assign the values if the from object has empty values? It seems like it does not

I think this is correct only for pointer types though. Non-pointers are [directly copied](https://github.com/crossplaneio/sample-stack-wordpress/pull/17/files#diff-a9ceb1ef9a800df85f5f80f38e4b6deeR88)

I just tested it and indeed the `Image string` field gets overridden with empty string from the actual `WordpressInstance` object.

Reply:
Dang! Ok. Thanks for testing it!